declarativeNetRequest's content rule list should not use the identifier for the filename.

xeenon opened the pull request:
#### c96350ffd7304e47f9995a186719057c6fb12a60
<pre>
declarativeNetRequest&apos;s content rule list should not use the identifier for the filename.
<a href="https://webkit.org/b/266198">https://webkit.org/b/266198</a>
<a href="https://rdar.apple.com/problem/119469953">rdar://problem/119469953</a>

Reviewed by Brian Weinstein and Alex Christensen.

Removes the legacy filename support from APIContentRuleListStore, per the comment that it can
be removed in 2022-2023.

Adds methods to allow compile, lookup, and remove with an explicit file path. This allows the
Web Extension dNR code to store the file at the path it desires, bypassing APIContentRuleListStore&apos;s
automatic file management of an explicit storage directory for rule lists. Web Extensions store their
rule list in a unique directory alongside other extension state files, so the rule list file name
does not need be unique or contain the rule list identifier.

* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::constructedPath): Drop legacy file name support. Added RELEASE_ASSERT for base.
(API::createExtension): Add some whitespace.
(API::getContentRuleListSourceFromMappedFile): Ditto.
(API::ContentRuleListStore::lookupContentRuleList): Call lookupContentRuleListFile.
(API::ContentRuleListStore::lookupContentRuleListFile): Added.
(API::ContentRuleListStore::getAvailableContentRuleListIdentifiers): Call compileContentRuleListFile.
(API::ContentRuleListStore::compileContentRuleList): Call compileContentRuleListFile.
(API::ContentRuleListStore::compileContentRuleListFile): Added.
(API::ContentRuleListStore::removeContentRuleList): Call removeContentRuleListFile.
(API::ContentRuleListStore::removeContentRuleListFile): Added.
(API::ContentRuleListStore::invalidateContentRuleListVersion): Drop false for legacy filename param.
(API::ContentRuleListStore::getContentRuleListSource): Drop legacy file name support.
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unloadDeclarativeNetRequestState):
(WebKit::WebExtensionContext::declarativeNetRequestContentRuleListFilePath): Added.
(WebKit::WebExtensionContext::addDeclarativeNetRequestRulesToPrivateUserContentControllers): Use API::ContentRuleListStore.
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules): Ditto.
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules): Ditto.
(WebKit::WebExtensionContext::declarativeNetRequestRuleStore): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm: Deleted ContentRuleList.LegacyVersionAndName.

Canonical link: <a href="https://commits.webkit.org/271903@main">https://commits.webkit.org/271903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d393d308fbaed4532c6a869bfbb4ca47151f544

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30022 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/8685 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/31333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/32522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/27136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/30682 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/10876 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/5937 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/32522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/30316 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/10876 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/31333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/32522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/10876 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/31333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/33859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/10876 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/31333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/33859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/6316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/5937 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/33859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/8060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/31333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/7065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/3873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -81,17 +81,15 @@ ContentRuleListStore::ContentRuleListStore(const WTF::String& storePath)
 
-// FIXME: Remove legacyFilename in 2022 or 2023 after users have had a chance to run the updating logic.
-static ASCIILiteral constructedPathPrefix(bool legacyFilename)
+static constexpr ASCIILiteral constructedPathPrefix()
```
This no longer needs to be a function.
static constexpr ASCIILiteral constructedPathPrefix { "ContentRuleList-" };

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -571,36 +563,43 @@ void ContentRuleListStore::compileContentRuleList(WTF::String&& identifier, WTF:
+    constexpr auto prefix = constructedPathPrefix();
+
+    for (const auto& fileName : listDirectory(m_storePath)) {
+        if (fileName.startsWith(prefix))
```
I don't think we should limit this to just files starting with the prefix.  If anything, we should append "ForTesting" to the name of this function.

xeenon replied:
I'll revert this.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -543,19 +531,23 @@ void ContentRuleListStore::getAvailableContentRuleListIdentifiers(CompletionHand
 
 void ContentRuleListStore::compileContentRuleList(WTF::String&& identifier, WTF::String&& json, CompletionHandler<void(RefPtr<API::ContentRuleList>, std::error_code)> completionHandler)
+{
+    compileContentRuleListFile(constructedPath(m_storePath, identifier), WTFMove(identifier), WTFMove(json), WTFMove(completionHandler));
```
This relies on undefined behavior.  It moves identifier and uses identifier as a function argument, both of which create function arguments.  This will only work on systems with calling conventions that have certain parameter evaluation order.

Instead, we need to call constructedPath and store the result, then call compileContentRuleListFile on the next line.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -464,55 +463,47 @@ static WTF::String getContentRuleListSourceFromMappedFile(const MappedData& mapp
 
 void ContentRuleListStore::lookupContentRuleList(WTF::String&& identifier, CompletionHandler<void(RefPtr<API::ContentRuleList>, std::error_code)> completionHandler)
+{
+    lookupContentRuleListFile(constructedPath(m_storePath, identifier), WTFMove(identifier), WTFMove(completionHandler));
```
Ditto below.  constructedPath result needs to be stored as a String before this call.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStore.mm:
```diff
@@ -151,6 +151,54 @@ - (void)removeContentRuleListForIdentifier:(NSString *)identifier completionHand
 
 @implementation WKContentRuleListStore (WKPrivate)
 
+- (void)_compileContentRuleListFileAtPath:(NSString *)path withIdentifier:(NSString *)identifier encodedContentRuleList:(NSString *)encodedContentRuleList completionHandler:(void (^)(WKContentRuleList *, NSError *))completionHandler
```
To match the rest of the ObjC API surface, the path needs to be an NSURL that we verify is a file URL with something like the existing function checkURLArgument.  That would mean this needs to be renamed to _compileContentRuleListAtFileURL.

Could you also add a check in WKContentRuleListStore.storeWithURL to verify it is a file URL?

It bothers me a bit that with this design you need to make a WKContentRuleListStore that has a m_storePath that you then don't use.  I think it would be better to just add SPI surface area that allows you to set your own file name that is in m_storePath.  Could we also add a check that the URL parameter here, after URLByDeletingLastPathComponent, is equal to a file URL constructed from m_storePath?  That would be stricter than just allowing any NSString.

achristensen07 replied:
Looking at your use of this, we don't actually need any ObjC SPI for this at all.  Just call API::ContentRuleListStore functions directly.  Can it be tested through WebExtension SPI/API instead of the API test you added?

achristensen07 commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2951,18 +2952,17 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 }
 
-WKContentRuleListStore *WebExtensionContext::declarativeNetRequestRuleStore()
+String WebExtensionContext::declarativeNetRequestContentRuleListFilePath()
```
It would probably be to our advantage to separate the extension parts of this from the non-extension parts of this into two PRs.

achristensen07 commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2951,18 +2952,17 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
-    auto contentBlockerStorePath = storageIsPersistent() ? storageDirectory() : String(FileSystem::createTemporaryDirectory(@"DeclarativeNetRequest"));
-    m_declarativeNetRequestRuleStore = [WKContentRuleListStore storeWithURL:[NSURL fileURLWithPath:contentBlockerStorePath]];
+    auto directoryPath = storageIsPersistent() ? storageDirectory() : String(FileSystem::createTemporaryDirectory(@"DeclarativeNetRequest"));
+    m_declarativeNetRequestContentRuleListFilePath = FileSystem::pathByAppendingComponent(directoryPath, "DeclarativeNetRequestContentRuleList.data"_s);
```
Is there existing data on people's disks that will be lost because of this file storage change?

xeenon replied:
Only the 3-4 of us working on the extensions code. I want to change this soon before broader adoption happens because of this.

achristensen07 commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestSQLiteStore.mm:
```diff
@@ -264,7 +264,7 @@ - (NSURL *)_databaseURL
     ASSERT([_storageType isEqualToString:@"dynamic"]);
 
-    NSString *databaseName = @"DynamicRules-DeclarativeNetRequest.db";
+    NSString *databaseName = @"DeclarativeNetRequestDynamicRules.db";
```
I don't understand what this change does.  Does it lose existing data?  Are the compiled content rule lists put into an sqlite database?

xeenon replied:
This is a separate DB for rules the extension can change dynamically. I'll split this out like you suggested above.

achristensen07 commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3062,7 +3061,7 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         if (!allJSONData.get().count) {
             removeDeclarativeNetRequestRules();
-            [declarativeNetRequestRuleStore() removeContentRuleListForIdentifier:uniqueIdentifier() completionHandler:^(NSError *) { }];
+            API::ContentRuleListStore::defaultStore().removeContentRuleListFile(declarativeNetRequestContentRuleListFilePath(), [](std::error_code) { });
```
The completion handler should probably be captured in the lambda and only be called with true if the error code is empty.  Otherwise we're assuming that something that is dispatched to a different thread will always and synchronously immediately succeed.

xeenon committed:
declarativeNetRequest's content rule list should not use the identifier for the filename.
https://webkit.org/b/266198
rdar://problem/119469953

Reviewed by Brian Weinstein and Alex Christensen.

Removes the legacy filename support from APIContentRuleListStore, per the comment that it can
be removed in 2022-2023.

Adds methods to allow compile, lookup, and remove with an explicit file path. This allows the
Web Extension dNR code to store the file at the path it desires, bypassing APIContentRuleListStore's
automatic file management of an explicit storage directory for rule lists. Web Extensions store their
rule list in a unique directory alongside other extension state files, so the rule list file name
does not need be unique or contain the rule list identifier.

* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::constructedPath): Drop legacy file name support. Added RELEASE_ASSERT for base.
(API::createExtension): Add some whitespace.
(API::getContentRuleListSourceFromMappedFile): Ditto.
(API::ContentRuleListStore::lookupContentRuleList): Call lookupContentRuleListFile.
(API::ContentRuleListStore::lookupContentRuleListFile): Added.
(API::ContentRuleListStore::getAvailableContentRuleListIdentifiers): Call compileContentRuleListFile.
(API::ContentRuleListStore::compileContentRuleList): Call compileContentRuleListFile.
(API::ContentRuleListStore::compileContentRuleListFile): Added.
(API::ContentRuleListStore::removeContentRuleList): Call removeContentRuleListFile.
(API::ContentRuleListStore::removeContentRuleListFile): Added.
(API::ContentRuleListStore::invalidateContentRuleListVersion): Drop false for legacy filename param.
(API::ContentRuleListStore::getContentRuleListSource): Drop legacy file name support.
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unloadDeclarativeNetRequestState):
(WebKit::WebExtensionContext::declarativeNetRequestContentRuleListFilePath): Added.
(WebKit::WebExtensionContext::addDeclarativeNetRequestRulesToPrivateUserContentControllers): Use API::ContentRuleListStore.
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules): Ditto.
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules): Ditto.
(WebKit::WebExtensionContext::declarativeNetRequestRuleStore): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm: Deleted ContentRuleList.LegacyVersionAndName.

Canonical link: https://commits.webkit.org/271903@main
